libraries/chain/finality_core.cpp

```diff
@@ -0,0 +1,338 @@
+#warning Remove undef NDEBUG for assert before RC
+//Undefine NDEBUG to enable assertions in CICD.
+#undef NDEBUG
+#include <cassert>
+
+#include <eosio/chain/finality_core.hpp>
+#include <eosio/chain/block_header.hpp>
+
+namespace eosio::chain {
+
+/**
+ *  @pre block_id is not null
+ *  @returns the extracted block_num from block_id
+ */
+block_num_type block_ref::block_num() const {
+   return block_header::num_from_id(block_id);
+}
+
+/**
+ *  @pre none
+ *
+ *  @post returned core has current_block_num() == block_num
+ *  @post returned core has latest_qc_claim() == {.block_num=block_num, .is_strong_qc=false}
+ *  @post returned core has final_on_strong_qc_block_num == block_num
+ *  @post returned core has last_final_block_num() == block_num
+ */
+finality_core finality_core::create_core_for_genesis_block(block_num_type block_num)
+{
+   return finality_core {
+      .links                        = {
+         qc_link{
+            .source_block_num = block_num,
+            .target_block_num = block_num,
+            .is_link_strong    = false,
+         },
+      },
+      .refs                         = {},
+      .final_on_strong_qc_block_num = block_num,
+   };
+
+   // Invariants 1 to 7 can be easily verified to be satisfied for the returned core.
+   // (And so, remaining invariants are also automatically satisfied.)
+}
+
+/**
+ *  @pre this->links.empty() == false
+ *  @post none
+ *  @returns block number of the core
+ */
+block_num_type finality_core::current_block_num() const
+{
+   assert(!links.empty()); // Satisfied by invariant 1.
+
+   return links.back().source_block_num;
+}
+
+/**
+ *  @pre this->links.empty() == false
+ *  @post none
+ *  @returns last final block_num in respect to the core
+ */
+block_num_type finality_core::last_final_block_num() const
+{
+   assert(!links.empty()); // Satisfied by invariant 1.
+
+   return links.front().target_block_num;
+}
+
+/**
+ *  @pre this->links.empty() == false
+ *  @post none
+ *  @returns latest qc_claim made by the core
+ */
+qc_claim finality_core::latest_qc_claim() const
+{
+   assert(!links.empty()); // Satisfied by invariant 1.
+
+   return qc_claim{.block_num = links.back().target_block_num, .is_strong_qc = links.back().is_link_strong};
+}
+
+/**
+ *  @pre last_final_block_num() <= block_num < current_block_num()
+ *
+ *  @post returned block_ref has block_num() == block_num
+ */
+const block_ref& finality_core::get_block_reference(block_num_type block_num) const
+{
+   assert(last_final_block_num() <= block_num); // Satisfied by precondition.
+   assert(block_num < current_block_num()); // Satisfied by precondition.
+
+   // If refs.empty() == true, then by invariant 3, current_block_num() == last_final_block_num(), 
+   // and therefore it is impossible to satisfy the precondition. So going forward, it is safe to assume refs.empty() == false.
+
+   const size_t ref_index = block_num - last_final_block_num();
```

Comment:
While it works, it is a little odd to use links here. Since we know the refs are monotonically increasing block numbers, I think the following is more natural:
```suggestion
   const size_t ref_index = block_num - refs.front().block_num();
```

Reply:
oh we cannot do that as refs can be empty while `last_final_block_num()` always returns a valid value.

Reply:
line above says: 
` So going forward, it is safe to assume refs.empty() == false.`